backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CpuPinningPolicyUnit.java

```diff
@@ -84,51 +94,46 @@ public List<VDS> filter(final SchedulingContext context,
                 boolean isDedicatedCpuPinningPossibleAtHost =
                         vdsCpuUnitPinningHelper.isExclusiveCpuPinningPossibleOnHost(
                                 vmToPendingDedicatedCpuPinnings,
-                                vm,
-                                host.getId());
-                if (isDedicatedCpuPinningPossibleAtHost) {
-                    candidates.add(host);
-                } else {
+                                exclusiveVms,
```

Comment:
Shouldn't we run the "preview pinning" for all vms? What if there is a mixture of manual and exclusive pinnings, we should make sure the dedicated knows about some of the CPUs being taken by manual pinning.

Reply:
The idea was: first do the dedicated, then do the shared.
I use the same instance of `cpuTopology` here. Therefore, once I run the `isExclusiveCpuPinningPossibleOnHost()` now, the dedicated are taken. I didn't think of shared pool as priority above to the exclusive policies.
Or, you mean there is a change a VM is scheduled on the VM (with manual pinning) and now we enter here before it is set?

Reply:
I think the manual pinning should be pinned first because it is more restrictive. Lets say we have a vmGroup with the following VMs:
- VM with manually pinned CPU 0#0
- VM with 1 dedicated CPU
If you run the dedicated CPU first, it could be pinned to host's Cpu 0 and that would make the manual pinning impossible and the VMs would not be scheduled even though the host could run them.

Maybe you should not divide the VMs to `exclusiveVms` and `sharedVms` here and rely on the sorting provided by the Scheduler manager (https://github.com/oVirt/ovirt-engine/pull/384). That way we can change the order of the VMs only on one place.

Reply:
OK, I changed placements, made `isExclusiveCpuPinningPossibleOnHost()` to run on all VMs when we have exclusive VMs. This, combining the VM sorting patch, should resolve it.
The plan now is to prioritize manual/resize and pin VMs, then isolate threads, then dedicated.